### PR TITLE
Developer's documentation

### PR DESCRIPTION
**Pull request by MaxThevenet**
This PR provides the first shot of developer's documentation. It contains:
- Some additional Doxygen documentation in the source code.
- A new section in the Sphinx documentation, for developers. This uses breathe to access Doxygen documentation from Sphinx.
- The whole Doxygen documentation is available on ReadTheDocs, converted to `.rst` format using the python module `exhale`.

This is actually explained in the doc itself, in the section `developers/documentation.rst`.

To install this doc on MacOS, Clone WarpX somewhere new, pull this branch and
```sh
cd Docs/
brew install doxygen
doxygen # compile the Doxygen documentation
pip install breathe exhale
make html # compile the Sphinx documentation, that includes the Doxygen doc thanks to breathe and exhale
```
Then you can look at the documentations with
```sh
open doxyhtml/index.html # Look at the Doxygen doc directly
open build/html/index.html # Look at the new Sphinx documentation, that includes the dev's doc
```

### CODE REVIEW

File: Docs/source/developers/amrex_basics.rst [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+WarpX is build on the Adaptive Mesh Refinement (AMR) library `AMReX <https://github.com/AMReX-Codes/amrex>`__. This section provides a very sporadic description of the main AMReX classes and concepts relevant for WarpX, that can serve as a reminder. Please read the AMReX basics `doc page <https://amrex-codes.github.io/amrex/docs_html/Basics.html>`__, of which this section is largely inspired.
+
+* ``amrex::Box``: Dimension-dependent lower and upper indices defining a rectangular volume in 3D (or surface in 2D) in the index space. ``Box`` is a lightweight meta-data class, with useful member functions.
+
+* ``amrex::BoxArray``: Collection of ``Box`` on a single AMR level. The information of which MPI rank owns which ``Box`` in a ``BoxArray`` is in ``DistributionMapping``.
+
+* ``amrex::FArrayBox``: Fortran-ordered array of floating-point ``amrex::Real`` elements defined on a ``Box``. A ``FArrayBox`` can represent scalar data or vector data, with ``ncomp`` components.
+
+* ``amrex::MultiFab``: Collection of `FAB` (= ``FArrayBox``) on a single AMR level. The concept of `ghost cells` is defined at the ``MultiFab`` level.
```
Comment (atmyers) [inline Docs/source/developers/amrex_basics.rst]: You might want to explicitly mention here that MultiFabs are distributed over MPI ranks.

File: Docs/source/developers/amrex_basics.rst [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+* ``amrex::Particle``: Data for 1 particle. It contains default member variables stored as Array-Of-Struct (AoS) (often called `data`, they are the 3D position, the particle ID and the index of the CPU owning the particle) and extra variables (often called `attribs`, in WarpX they are the momentum, field on particle etc.), stored at Struct-Of-Arrays (SoA).
```
Comment (atmyers) [inline Docs/source/developers/amrex_basics.rst]: the class called `amrex::Particle` only has the AoS data, so this is slightly misleading. I'd merge the above description of AoS versus SoA in at the end of the next section about `ParticleContainer` and skip documenting `amrex::Particle`.

File: Docs/source/developers/amrex_basics.rst (hunk 1 above)
Comment (MaxThevenet) [reply Docs/source/developers/amrex_basics.rst]: Thanks, that was fixed in the last commit

File: Docs/source/developers/amrex_basics.rst (hunk 2 above)
Comment (MaxThevenet) [reply Docs/source/developers/amrex_basics.rst]: Thanks, that was fixed in the last commit

File: Docs/source/conf.py [hunk 3]
```diff
@@ -162,3 +164,34 @@
 
 # Example configuration for intersphinx: refer to the Python standard library.
 intersphinx_mapping = {'https://amrex-codes.github.io/': None}
+
+# Setup the breathe extension
+breathe_projects = {
+    "My Project": "../doxyxml/"
```
Comment (ax3l) [inline Docs/source/conf.py]:
```suggestion
    "WarpX": "../doxyxml/"
```

File: Docs/source/conf.py [hunk 4]
```diff
@@ -162,3 +164,34 @@
...
+}
+breathe_default_project = "My Project"
```
Comment (ax3l) [inline Docs/source/conf.py]:
```suggestion
breathe_default_project = "WarpX"
```

File: Docs/source/developers/amrex_basics.rst [hunk 5]
```diff
@@ -0,0 +1,26 @@
+AMReX basics (excessively basic)
+================================
+
+WarpX is build on the Adaptive Mesh Refinement (AMR) library `AMReX <https://github.com/AMReX-Codes/amrex>`__. This section provides a very sporadic description of the main AMReX classes and concepts relevant for WarpX, that can serve as a reminder. Please read the AMReX basics `doc page <https://amrex-codes.github.io/amrex/docs_html/Basics.html>`__, of which this section is largely inspired.
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]: Is there a difference between one and two underscores in links?

File: Docs/source/developers/amrex_basics.rst [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+* ``amrex::FArrayBox``: Fortran-ordered array of floating-point ``amrex::Real`` elements defined on a ``Box``. A ``FArrayBox`` can represent scalar data or vector data, with ``ncomp`` components.
+
+* ``amrex::MultiFab``: Collection of `FAB` (= ``FArrayBox``) on a single AMR level, distributed over MPI ranks. The concept of `ghost cells` is defined at the ``MultiFab`` level.
+
+* ``amrex::ParticleContainer``: A collection of particles, typically for particles of a physical species. Particles in a ``ParticleContainer`` are organized per ``Box``. Particles in a ``Box`` are organized per tile (this feature is off when running on GPU). Particles within a tile are stored in several structures, each being contiguous in memory: (i) an Array-Of-Struct (AoS) (often called `data`, they are the 3D position, the particle ID and the index of the CPU owning the particle), where the Struct is an ``amrex::Particle`` and (ii) Struct-Of-Arrays (SoA) for extra variables (often called `attribs`, in WarpX they are the momentum, field on particle etc.).
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]:
```suggestion
* ``amrex::ParticleContainer``: A collection of particles, typically for particles of a physical species. Particles in a ``ParticleContainer`` are organized per ``Box``. Particles in a ``Box`` are organized per tile (this feature is off when running on GPU). Particles within a tile are stored in several structures, each being contiguous in memory: (i) an Array-Of-Struct (AoS) (often called `data`, they are the 3D position, the particle ID and the index of the CPU owning the particle), where the Struct is an ``amrex::Particle`` and (ii) Struct-Of-Arrays (SoA) for extra variables (often called ``attribs``, in WarpX they are the momentum, field on particle etc.).
```

File: Docs/source/developers/amrex_basics.rst [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+The simulation domain is decomposed in several ``Box``, and each MPI rank owns (and performs operations on) the fields and particles defined on a few of these ``Box``, but has the metadata of all of them. For convenience, AMReX provides iterators, to easily iterate over all ``FArrayBox`` (or even tile-by-tile, optionally) in a ``MultiFab`` own by the MPI rank (``MFIter``), or over all particles in a ``ParticleContainer`` on a per-box basis (``ParIter``, or its derived class ``WarpXParIter``). These are respectively done in loops like::
+
+  // mf is a pointer to MultiFab
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]: maybe a
```
.. code-block:: cpp
```

before such blocks?
```suggestion
.. code-block:: cpp

   // mf is a pointer to MultiFab
```

File: Docs/source/developers/diagnostics.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+===========
+
+Regular diagnostics (plotfiles)
+-------------------------------
+
+.. note::
+
+   Section empty!
+
+Back-transformed-diagnostics
```
Comment (ax3l) [inline Docs/source/developers/diagnostics.rst]:
```suggestion
Back-Transformed Diagnostics
```

File: Docs/source/developers/diagnostics.rst [hunk 9]
```diff
@@ -0,0 +1,16 @@
+Diagnostics
+===========
+
+Regular diagnostics (plotfiles)
```
Comment (ax3l) [inline Docs/source/developers/diagnostics.rst]:
```suggestion
Regular Diagnostics (plotfiles)
```

File: Docs/source/developers/documentation.rst [hunk 10]
```diff
@@ -0,0 +1,41 @@
+Documentation
+=============
+
+.. warning::
+   Needs info on how to install all pieces to compile the doc! Some basic info is in the CONTRIBUTING.md
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]: I think this needs a newline.
```suggestion

   Needs info on how to install all pieces to compile the doc! Some basic info is in the ``CONTRIBUTING.md`` .
```

File: Docs/source/developers/documentation.rst [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+Doxygen documentation
+---------------------
+
+WarpX uses a Doxygen documentation. Whenever you create a new class, please document it where it is declared (typically in the header file)::
+
+   /**
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
.. code-block:: cpp

   /**
```

File: Docs/source/developers/amrex_basics.rst [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  for ( amrex::MFIter mfi(mf, false); mfi.isValid(); ++mfi ) { ... }
+
+and::
+
+  // *this is a pointer to a ParticleContainer
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]:
```suggestion
.. code-block:: cpp

   // *this is a pointer to a ParticleContainer
```

File: Docs/source/developers/amrex_basics.rst [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+The simulation domain is decomposed in several ``Box``, and each MPI rank owns (and performs operations on) the fields and particles defined on a few of these ``Box``, but has the metadata of all of them. For convenience, AMReX provides iterators, to easily iterate over all ``FArrayBox`` (or even tile-by-tile, optionally) in a ``MultiFab`` own by the MPI rank (``MFIter``), or over all particles in a ``ParticleContainer`` on a per-box basis (``ParIter``, or its derived class ``WarpXParIter``). These are respectively done in loops like::
+
+  // mf is a pointer to MultiFab
+  for ( amrex::MFIter mfi(mf, false); mfi.isValid(); ++mfi ) { ... }
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]:
```suggestion
   for ( amrex::MFIter mfi(mf, false); mfi.isValid(); ++mfi ) { ... }
```

File: Docs/source/developers/amrex_basics.rst [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+and::
+
+  // *this is a pointer to a ParticleContainer
+  for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti) { ... }
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]:
```suggestion
   for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti) { ... }
```

File: Docs/source/developers/documentation.rst [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    *
+    * If you are kind enough, also quickly explain how things in my_class work.
+    * (typically a few more lines)
+    */
+   class my_class
+   { ... }
+
+Doxygen reads this docstring, so please be accurate with the syntax! See `Doxygen manual <http://www.doxygen.nl/manual/docblocks.html>`__ for more information. Similarly, please document functions when you declare them (typically in a header file) like::
+
+  /**
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
.. code-block:: cpp

  /**
```

File: Docs/source/developers/documentation.rst [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+   /**
+    * few-line description explaining the purpose of my_class.
+    *
+    * If you are kind enough, also quickly explain how things in my_class work.
+    * (typically a few more lines)
+    */
+   class my_class
+   { ... }
+
+Doxygen reads this docstring, so please be accurate with the syntax! See `Doxygen manual <http://www.doxygen.nl/manual/docblocks.html>`__ for more information. Similarly, please document functions when you declare them (typically in a header file) like::
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
Doxygen reads this docstring, so please be accurate with the syntax! See `Doxygen manual <http://www.doxygen.nl/manual/docblocks.html>`__ for more information. Similarly, please document functions when you declare them (typically in a header file) like:
```

File: Docs/source/developers/documentation.rst [hunk 17]
```diff
@@ -0,0 +1,41 @@
+Documentation
+=============
+
+.. warning::
+   Needs info on how to install all pieces to compile the doc! Some basic info is in the CONTRIBUTING.md
+
+Doxygen documentation
+---------------------
+
+WarpX uses a Doxygen documentation. Whenever you create a new class, please document it where it is declared (typically in the header file)::
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]: Not sure if a typo or syntax :D 
```suggestion
WarpX uses a Doxygen documentation. Whenever you create a new class, please document it where it is declared (typically in the header file):
```

File: Docs/source/developers/documentation.rst [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+   *
+   * \param[in,out] my_int a pointer to an integer variable on which
+   *                       my_function will operate.
+   */
+  void my_class::my_function(int* my_int);
+
+Breathe documentation
+---------------------
+
+Your Doxygen documentation is not only useful for people looking into the code, it is also part of the `WarpX online documentation <https://ecp-warpx.github.io>`__ based on `Sphinx <http://www.sphinx-doc.org/en/master/>`__! This is done using the Python module `Breathe <http://breathe.readthedocs.org>`__, that allows you to read Doxygen documentation dorectly in the source and include it in your Sphinx documentation, by calling Breathe functions. For instance, the following line will get the Doxygen documentation for ``WarpXParticleContainer`` in ``Source/Particles/WarpXParticleContainer.H`` and include it to the html page generated by Sphinx::
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
Your Doxygen documentation is not only useful for people looking into the code, it is also part of the `WarpX online documentation <https://ecp-warpx.github.io>`__ based on `Sphinx <http://www.sphinx-doc.org/en/master/>`__! This is done using the Python module `Breathe <http://breathe.readthedocs.org>`__, that allows you to read Doxygen documentation dorectly in the source and include it in your Sphinx documentation, by calling Breathe functions. For instance, the following line will get the Doxygen documentation for ``WarpXParticleContainer`` in ``Source/Particles/WarpXParticleContainer.H`` and include it to the html page generated by Sphinx:
```
?

File: Docs/source/developers/documentation.rst [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+  .. doxygenclass:: WarpXParticleContainer
+
+Exhale documentation?
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
Exhale documentation
```

File: Docs/source/developers/documentation.rst [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+---------------------
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
--------------------
```

File: Docs/source/developers/documentation.rst [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+
+Very similar to Breathe, the Python module `exhale <https://exhale.readthedocs.io/en/latest/>`__ reads the full Doxygen documentation and renders it in `rst` format, and is accessible from the main WarpX ReadTheDocs page.
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]:
```suggestion
Very similar to Breathe, the Python module `exhale <https://exhale.readthedocs.io/en/latest/>`__ reads the full Doxygen documentation and renders it in ` rst <https://en.wikipedia.org/wiki/ReStructuredText>`__ format, and is accessible from the main WarpX ReadTheDocs page.
```

File: Docs/source/developers/fields.rst [hunk 22]
```diff
@@ -0,0 +1,121 @@
+Fields
+======
+
+.. note::
+
+   Add info on staggering and domain decomposition. Synchronize with section ``initialization``.
+
+The main fields are the electric field ``Efield``, the magnetic field ``Bfield``, the current density ``current`` and the charge density ``rho``. When a divergence-cleaner is used, another field ``F`` (containing ``div(Efield)-rho``).
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
The main fields are the electric field ``Efield``, the magnetic field ``Bfield``, the current density ``current`` and the charge density ``rho``. When a divergence-cleaner is used, we add another field ``F`` (containing :math:`\vec \nabla \vec E - \rho`).
```

File: Docs/source/developers/amrex_basics.rst [hunk 23]
```diff
@@ -0,0 +1,26 @@
+AMReX basics (excessively basic)
```
Comment (ax3l) [inline Docs/source/developers/amrex_basics.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-amrex-basics:

AMReX basics (excessively basic)
```

File: Docs/source/developers/developers.rst [hunk 24]
```diff
@@ -0,0 +1,18 @@
+Developers documentation
```
Comment (ax3l) [inline Docs/source/developers/developers.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers:

Developers documentation
```

File: Docs/source/developers/diagnostics.rst [hunk 25]
```diff
@@ -0,0 +1,16 @@
+Diagnostics
```
Comment (ax3l) [inline Docs/source/developers/diagnostics.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-diagnostics:

Diagnostics
```

File: Docs/source/developers/documentation.rst [hunk 26]
```diff
@@ -0,0 +1,41 @@
+Documentation
```
Comment (ax3l) [inline Docs/source/developers/documentation.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-docs:

Documentation
```

File: Docs/source/developers/fields.rst [hunk 27]
```diff
@@ -0,0 +1,121 @@
+Fields
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-fields:

Fields
```

File: Docs/source/developers/initialization.rst [hunk 28]
```diff
@@ -0,0 +1,21 @@
+Initialization
```
Comment (ax3l) [inline Docs/source/developers/initialization.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-initialization:

Initialization
```

File: Docs/source/developers/moving_window.rst [hunk 29]
```diff
@@ -0,0 +1,6 @@
+Moving Window
```
Comment (ax3l) [inline Docs/source/developers/moving_window.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-moving-window:

Moving Window
```

File: Docs/source/developers/particles.rst [hunk 30]
```diff
@@ -0,0 +1,98 @@
+Particles
```
Comment (ax3l) [inline Docs/source/developers/particles.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-particles:

Particles
```

File: Docs/source/developers/portability.rst [hunk 31]
```diff
@@ -0,0 +1,5 @@
+Portability
```
Comment (ax3l) [inline Docs/source/developers/portability.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-portability:

Portability
```

File: Docs/source/developers/profiling.rst [hunk 32]
```diff
@@ -0,0 +1,5 @@
+Profiling
```
Comment (ax3l) [inline Docs/source/developers/profiling.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-profiling:

Profiling
```

File: Docs/source/developers/repo_organization.rst [hunk 33]
```diff
@@ -0,0 +1,22 @@
+WarpX structure
```
Comment (ax3l) [inline Docs/source/developers/repo_organization.rst]: Let's add a label here for referencing in text:
```suggestion
.. _developers-repo-structure:

WarpX structure
```

File: Docs/source/developers/fields.rst [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+======
+
+.. note::
+
+   Add info on staggering and domain decomposition. Synchronize with section ``initialization``.
+
+The main fields are the electric field ``Efield``, the magnetic field ``Bfield``, the current density ``current`` and the charge density ``rho``. When a divergence-cleaner is used, another field ``F`` (containing ``div(Efield)-rho``).
+
+Due the AMR strategy used in WarpX (see section :doc:`../../theory/amr` for a complete description), each field on a given refinement level ``lev`` (except for the coarsest `0`) is defined on:
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]: You can add a section reference as indicated in the other files, e.g. call it `theory-amr`, and then link to it via:
```suggestion
Due the AMR strategy used in WarpX (see section :ref:`Theory: AMR <theory-amr>` for a complete description), each field on a given refinement level ``lev`` (except for the coarsest ``0``) is defined on:
```

File: Docs/source/developers/fields.rst [hunk 35]
```diff
@@ -0,0 +1,121 @@
...
+
+* **the fine patch** (suffix ``_fp``, the actual resolution on ``lev``).
+
+* **the coarse patch** (suffix ``_cp``, same physical domain with the resolution of MR level ``lev-1``).
+
+* **the auxiliary grid** (suffix ``_aux``, same resolution as ``_fp``), from which the fields are gathered from the grids to particle positions. For this reason. only `E` and `B` are defined on this ``_aux`` grid (not the current density or charge density).
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
* **the auxiliary grid** (suffix ``_aux``, same resolution as ``_fp``), from which the fields are gathered from the grids to particle positions. For this reason. only ``E`` and ``B`` are defined on this ``_aux`` grid (not the current density or charge density).
```

File: Docs/source/developers/fields.rst [hunk 36]
```diff
@@ -0,0 +1,121 @@
...
+
+* In some conditions, i.e., when buffers are used for the field gather (for numerical reasons), a **copy of E and B on the auxiliary grid** ``_aux`` **of the  level below** ``lev-1`` is stored in fields with suffix ``_cax`` (for `coarse aux`).
+
+As an example, the structures for the electric field are ``Efield_fp``, ``Efield_cp``, ``Efield_aux`` (and optionally ``Efield_cax``).
+
+Declaration
+-----------
+
+All the fields described above are public members of class ``WarpX``, defined in ``WarpX.H``. They are defined as an ``amrex::Vector`` (over MR levels) of ``std::array`` (for the 3 spatial components `Ex`, `Ey`, `Ez`) of ``std::unique_ptr`` of ``amrex::MultiFab``, i.e.,::
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
All the fields described above are public members of class ``WarpX``, defined in ``WarpX.H``. They are defined as an ``amrex::Vector`` (over MR levels) of ``std::array`` (for the 3 spatial components :math:`E_x`, :math:`E_y`, :math:`E_z`) of ``std::unique_ptr`` of ``amrex::MultiFab``, i.e.:
```

File: Docs/source/developers/fields.rst [hunk 37]
```diff
@@ -0,0 +1,121 @@
...
+
+  amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > > Efield_fp;
+
+Hence, `Ex` on MR level ``lev`` is a pointer to an ``amrex::MultiFab``. The other fields are organized in the same way.
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
Hence, ``Ex`` on MR level ``lev`` is a pointer to an ``amrex::MultiFab``. The other fields are organized in the same way.
```

File: Docs/source/developers/fields.rst [hunk 38]
```diff
@@ -0,0 +1,121 @@
...
+
+Allocation and initialization
+-----------------------------
+
+The ``MultiFab`` constructor (for, e.g., `Ex` on level `lev`) is called in ``WarpX::AllocLevelMFs``.
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
The ``MultiFab`` constructor (for, e.g., ``Ex`` on level ``lev``) is called in ``WarpX::AllocLevelMFs``.
```

File: Docs/source/developers/fields.rst [hunk 39]
```diff
@@ -0,0 +1,121 @@
...
+
+Field solver
+------------
+
+The field solver is performed in ``WarpX::EvolveE`` for the electric field and ``WarpX::EvolveB`` for the magnetic field, called from ``WarpX::OneStep_nosub`` in ``WarpX::EvolveEM``. This section describes the FDTD field push (the PSATD field push should be added later). It is implemented in ``Source/FieldSolver/WarpXPushFieldsEM.cpp``.
+
+As all cell-wise operation, the field push is done as follows (this is split in multiple functions in the actual implementation to aboid code duplication)
+::
+
+   // Loop over MR levels
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
.. code-block:: cpp

   // Loop over MR levels
```

File: Docs/source/developers/fields.rst [hunk 40]
```diff
@@ -0,0 +1,121 @@
...
+      MultiFab* Ex = Efield_fp[lev][0].get();
+      // Loop over boxes (or tiles if not on GPU)
+      for ( MFIter mfi(*Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+          // Apply field solver on the FAB
+      }
+  }
+
+The innermost step ``// Apply field solver on the FAB`` could be done with 3 nested ``for`` loops for the 3 dimensions (in 3D). However, for portability reasons (see section :doc:`./portability`), this is done in two steps: (i) extract AMReX data structures into plain-old-data simple structures, and (ii) call a general ``ParallelFor`` function (translated into nested loops on CPU or a kernel launch on GPU, for instance)::
+
+  // Get Box corresponding to the current MFIter
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
.. code-block:: cpp

  // Get Box corresponding to the current MFIter
```

File: Docs/source/developers/fields.rst [hunk 41]
```diff
@@ -0,0 +1,121 @@
...
+   for (int lev = 0; lev <= finest_level; ++lev) {
+      // Get pointer to MultiFab Ex on level lev
+      MultiFab* Ex = Efield_fp[lev][0].get();
+      // Loop over boxes (or tiles if not on GPU)
+      for ( MFIter mfi(*Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+          // Apply field solver on the FAB
+      }
+  }
+
+The innermost step ``// Apply field solver on the FAB`` could be done with 3 nested ``for`` loops for the 3 dimensions (in 3D). However, for portability reasons (see section :doc:`./portability`), this is done in two steps: (i) extract AMReX data structures into plain-old-data simple structures, and (ii) call a general ``ParallelFor`` function (translated into nested loops on CPU or a kernel launch on GPU, for instance)::
```
Comment (ax3l) [inline Docs/source/developers/fields.rst]:
```suggestion
The innermost step ``// Apply field solver on the FAB`` could be done with 3 nested ``for`` loops for the 3 dimensions (in 3D). However, for portability reasons (see section :ref:`Developers: Portability <developers-portability>`), this is done in two steps: (i) extract AMReX data structures into plain-old-data simple structures, and (ii) call a general ``ParallelFor`` function (translated into nested loops on CPU or a kernel launch on GPU, for instance)::
```

File: Docs/source/developers/particles.rst [hunk 42]
```diff
@@ -0,0 +1,98 @@
...
+Finally, all particle species (physical plasma species ``PhysicalParticleContainer``, photon species ``PhotonParticleContainer`` or non-physical species ``LaserParticleContainer``) are stored in ``MultiParticleContainer``. The class ``WarpX`` holds one instance of ``MultiParticleContainer`` as a member variable, called ``WarpX::mypc`` (where `mypc` stands for "my particle containers"):
+
+.. doxygenclass:: MultiParticleContainer
+
+Loop over particles
+-------------------
+
+A typical loop over particles reads::
+
+  // pc is a std::unique_ptr<WarpXParticleContainer>
```
Comment (ax3l) [inline Docs/source/developers/particles.rst]:
```suggestion
.. code-block:: cpp

  // pc is a std::unique_ptr<WarpXParticleContainer>
```

File: Docs/source/developers/particles.rst [hunk 43]
```diff
@@ -0,0 +1,98 @@
...
+  const long np = pti.numParticles();
+
+Link fields and particles?
+--------------------------
+
+In WarpX, the loop over boxes through a ``MultiFab`` iterator ``MFIter`` and the loop over boxes through a ``ParticleContainer`` iterator ``WarpXParIter`` are consistent.
+
+On a loop over boxes in a ``MultiFab`` (``MFIter``), it can be useful to access particle data on a GPU-friendly way. This can be done by::
+
+  // Index of grid (= box)
```
Comment (ax3l) [inline Docs/source/developers/particles.rst]:
```suggestion
.. code-block:: cpp

  // Index of grid (= box)
```

File: Docs/source/developers/particles.rst [hunk 44]
```diff
@@ -0,0 +1,98 @@
...
+  auto& ptile = GetParticles(lev)[std::make_pair(grid_id,tile_id)];
+  ParticleType* pp = particle_tile.GetArrayOfStructs()().data();
+  // Only need attribs (i.e., SoA data)
+  auto& soa = ptile.GetStructOfArrays();
+  // As an example, let's get the ux momentum
+  const ParticleReal * const AMREX_RESTRICT ux = soa.GetRealData(PIdx::ux).data();
+
+On a loop over particles it can be useful to access the fields on the box we are looping over (typically when we use both field and particle data on the same box, for field gather or current deposition for instance). This is done for instance by adding this snippet in ``[MY INNER LOOP]``::
+
+  // E is a reference to, say, WarpX::Efield_aux
```
Comment (ax3l) [inline Docs/source/developers/particles.rst]:
```suggestion
.. code-block:: cpp

  // E is a reference to, say, WarpX::Efield_aux
```

File: Docs/source/developers/amrex_basics.rst (hunk 5 above)
Comment (MaxThevenet) [reply Docs/source/developers/amrex_basics.rst]: There seems to be, but that's still not very clear to me 
https://stackoverflow.com/questions/27420317/restructured-text-rst-http-links-underscore-vs-use

Comment (atmyers) [inline Docs/source/developers/amrex_basics.rst]: is *built on

**Comment by MaxThevenet** [comment]
Thanks @ax3l and @atmyers for your suggestions!! I incorporated the in the last commits.
@ax3l
```rst
blah-blah::

   foo
```
is the same as
```rst
blah-blah

::

   foo
```
I think you were wondering about that.

File: Docs/requirements.txt [hunk 45]
```diff
@@ -2,3 +2,4 @@ sphinx_rtd_theme>=0.3.1
 recommonmark
 sphinx
 pygments
+exhale
```
Comment (ax3l) [inline Docs/requirements.txt]:
```suggestion
breathe
exhale
```

**Pull request merged**